birdset/modules/models/ssast.py

```diff
@@ -0,0 +1,462 @@
+            self.original_embedding_dim = self.v.pos_embed.shape[2]
+
+            # SSL Pretraining Code
+            self.softmax = nn.Softmax(dim=-1)
+            self.lsoftmax = nn.LogSoftmax(dim=-1)
+            self.fshape, self.tshape = fshape, tshape
+            self.fstride, self.tstride = fstride, tstride
+            self.input_fdim, self.input_tdim = input_fdim, input_tdim
+            # this is a trick to make state_dict to track pretraining input_fdim and input_tdim and save them by using torch.save
+            self.p_input_fdim, self.p_input_tdim = nn.Parameter(torch.tensor(input_fdim), requires_grad=False), nn.Parameter(torch.tensor(input_tdim), requires_grad=False)
+
+            # masked patch classification (discriminative objective) layer
+            # we use two layers for pretext task, but using a single layer has similar performance.
+            # we map the output of transformer (768-dim for base models) to 256-dim patch input space, and then dot product with flattened patch input (also 256-dim) to calculate loss.
+            # alternatively, you can map the output of transformer to 768-dim patch embedding space, and dot product with patch embedding. Performance-wise they are similar, but map to 256 space is more efficient.
+            self.cpredlayer = nn.Sequential(nn.Linear(self.original_embedding_dim, self.original_embedding_dim), nn.ReLU(), nn.Linear(self.original_embedding_dim, 256))
+            # masked patch reconstruction (generative objective) layer
+            self.gpredlayer = nn.Sequential(nn.Linear(self.original_embedding_dim, self.original_embedding_dim), nn.ReLU(), nn.Linear(self.original_embedding_dim, 256))
+            self.unfold = torch.nn.Unfold(kernel_size=(fshape, tshape), stride=(fstride, tstride))
+
+            # we use learnable mask embedding (follow the BEIT paper), but using a fixed mask embedding (e.g., 0) leads to same performance.
+            self.mask_embed = nn.Parameter(torch.zeros([1, 1, self.original_embedding_dim]))
+            self.mask_embed = torch.nn.init.xavier_normal_(self.mask_embed)
+
+            # get the intermediate shape
+            self.p_f_dim, self.p_t_dim = self.get_shape(fstride, tstride, input_fdim, input_tdim, fshape, tshape)
+            num_patches = self.p_f_dim * self.p_t_dim
+            self.num_patches = num_patches
+            self.v.patch_embed.num_patches = num_patches
+            print('pretraining patch split stride: frequency={:d}, time={:d}'.format(fstride, tstride))
+            print('pretraining patch shape: frequency={:d}, time={:d}'.format(fshape, tshape))
+            print('pretraining patch array dimension: frequency={:d}, time={:d}'.format(self.p_f_dim, self.p_t_dim))
+            print('pretraining number of patches={:d}'.format(num_patches))
+
+            # the linear patch projection layer, use 1 channel for spectrogram rather than the original 3 channels for RGB images.
```

Comment:
@XgamerTV Was SSAST trained on a 3 dimensional specrogram? If so, we probably have to do the same preprocessing to use the pretrained model.